File: simple-node-ai-clients/nearest.js
const frame = require('./_frame');

let fired;

frame((state, obj) => {
  console.log(state.radar);
  if (!fired) {
    fired = {};
    state.myTank.forEach(tank => {
      fired[tank.id] = 0;
    });
  }
  state.myTank.forEach(tank => {
    let fire = state.radar.fire[tank.id];
    fire = [fire.up, fire.left, fire.down, fire.right].filter(v => v);
    if (fire.length > 0) {
      obj[tank.id] = fire[0].action;
      return;
    }
    let dist = 100;
    let target = null;
    state.enemyTank.forEach(enemy => {
      let d = Math.abs(enemy.x - tank.x) + Math.abs(enemy.y - tank.y);
      if (d < dist) {
        dist = d;
        target = enemy;
      }
    });
    if (!target) {
      return;
    }
    // if (fired[tank.id] == 0) {
    //   if (target.x == tank.x) {
    //     if (Math.abs(target.y - tank.y) < 10) {
    //       fired[tank.id] = 4;
    //       obj[tank.id] = {
    //         action: target.y < tank.y ? 'fire-up' : 'fire-down',
    //       };
    //       return;
    //     }
    //   } else if (target.y == tank.y) {
    //     if (Math.abs(target.x - tank.x) < 10) {
    //       fired[tank.id] = 4;
    //       obj[tank.id] = {
    //         action: target.x < tank.x ? 'fire-left' : 'fire-right',
    //       };
    //       return;
    //     }
    //   }
    // } else {
    //   fired[tank.id]--;
    // }
    obj[tank.id] = {
      // x: Math.floor(state.terain[0].length / 2),
      // y: Math.floor(state.terain.length / 2),
      x: target.x,
      y: target.y,
      action: 'travel-with-dodge',
    };
  });
});

File: go-ai/src/tactics/nearest.go
package tactics

import (
	f "framework"
	"math/rand"
)

type Nearest struct {
	round int
}

func NewNearest() *Nearest {
	return &Nearest { round: 0 }
}

func (self *Nearest) Init(state *f.GameState) {
}

func (self *Nearest) Plan(state *f.GameState, radar *f.RadarResult, objective map[string]f.Objective) {
	self.round++
	// if self.round < 5 {
	// 	return
	// }
	tankloop: for _, tank := range state.MyTank {
		fireRadar := radar.Fire[tank.Id]
		faith := 0.
		var pfire *f.RadarFire
		for _, fire := range []*f.RadarFire { fireRadar.Up, fireRadar.Down, fireRadar.Left, fireRadar.Right } {
			if fire != nil && fire.Sin < 0.5 && fire.Faith > faith {
				pfire = fire
			}
		}
		if pfire != nil {
			objective[tank.Id] = f.Objective {
				Action: pfire.Action,
			}
			continue tankloop
		}

		least := 99999
		var ttank *f.Tank
		for _, etank := range state.EnemyTank {
			dist := abs(tank.Pos.X - etank.Pos.X) + abs(tank.Pos.Y - etank.Pos.Y)
			if dist < least {
				ttank = &etank
				least = dist
			}
		}
		if ttank != nil {
			travel := f.ActionTravel
			_ = rand.Int()
			// if radar.Dodge[tank.Id].Threat > 0.9 && rand.Int() % 3 > 0 {
			// 	travel = f.ActionTravelWithDodge
			// }
			p := ttank.Pos
			switch p.Direction {
			case f.DirectionUp:
				p.Y -= state.Params.TankSpeed
			case f.DirectionDown:
				p.Y += state.Params.TankSpeed
			case f.DirectionLeft:
				p.X -= state.Params.TankSpeed
			case f.DirectionRight:
				p.X += state.Params.TankSpeed
			}
			objective[tank.Id] = f.Objective {
				Action: travel,
				Target: p,
				// Target: f.Position { X: 15, Y: 15 },
			}
		}
	}
}

func (self *Nearest) End(state *f.GameState) {
}

func abs (val int) int {
	if val < 0 {
		return -val
	} else {
		return val
	}
}
